mingwrt/man/wcrtomb.3.man
.\" vim: ft=nroff
.TH %PAGEREF% MinGW "MinGW Programmer's Reference Manual"
.
.SH NAME
.B \%wcrtomb
\- convert a wide character to a multibyte sequence
.
.
.SH SYNOPSIS
.B  #include
.RB < wchar.h >
.PP
.B  size_t wcrtomb( char
.BI * s ,
.B  wchar_t
.IB wc ,
.B  mbstate_t
.BI * ps
.B  );
.
.
.SH DESCRIPTION
The
.BR \%wcrtomb ()
function determines the number of bytes which are required,
starting from the conversion state represented by the
.B \%mbstate_t
object at
.IR *ps ,
to accommodate the multibyte character sequence,
in the codeset associated with the
.B \%LC_CTYPE
category of the active process locale,
which represents the completed conversion of
the wide character specified by
.IR wc .
.
.PP
In the special case,
when
.I s
is a NULL pointer,
the
.I wc
argument is ignored,
and the call is evaluated as if it had been invoked as
.PP
.RS 4n
.EX
wcrtomb( buf, L'\e0', ps )
.EE
.RE
.PP
returning the effect of conversion of the NUL wide character,
as a completion of any intermediate conversion state specified in
.IR *ps ,
but without storing the converted multibyte sequence;
(in this special case,
the
.B \%ISO\(hyC99
standard specifies that
.I buf
should be an internal buffer,
but since such a buffer becomes effectively inaccessible,
storage of any converted multibyte sequence is unnecessary).
.
.PP
Conversely,
in the normal case,
when
.I s
is not a NULL pointer,
the
.BR \%wcrtomb ()
function converts the wide character,
represented by
.IR wc ,
to the corresponding multibyte character sequence,
which is stored in the byte array starting at
.IR *s ,
and the function return value is set to
the number of bytes stored.
.
.
.SH RETURN VALUE
When conversion is successful,
regardless of whether the resultant multibyte sequence is stored,
or not,
the
.BR wcrtomb ()
function returns the number of bytes which are,
or which would be,
stored at
.IR *s .
.
.PP
If the result of conversion represents a completed multibyte sequence,
the conversion state,
represented by
.IR *ps ,
is updated to represent the
.I initial
.IR state .
Conversely,
if the result of conversion is equivalent to the conversion of a
.I high
.IR surrogate ,
nothing is stored,
the return value is set to
.IR zero ,
and the conversion state is updated to represent a pending
.I surrogate pair
completion.
.
.
.SH ERROR CONDITIONS
If the wide character,
passed as
.IR wc ,
either cannot be converted to a valid multibyte sequence,
or does not complete a pending
.I surrogate pair
which can be represented as a valid multibyte sequence,
in the codeset of the active
.B \%LC_CTYPE
locale category,
.I \%errno
is set to
.BR \%EILSEQ ,
the
.BR wcrtomb ()
function returns
.IR (size_t)(\-1) ,
and the conversion state is unspecified.
.
.
.SH STANDARDS CONFORMANCE
Except in respect of its extended provision for handling of
.IR surrogate\ pairs ,
and to the extent that it may be affected by limitations
of the underlying \%MS\(hyWindows API,
the
.I \%libmingwex
implementation of
.BR \%wcrtomb ()
conforms generally to
.BR \%ISO\(hyC99 ,
.BR \%POSIX.1\(hy2001 ,
and
.BR \%POSIX.1\(hy2008 .
.
.
.\"SH EXAMPLE
.
.
.SH CAVEATS AND BUGS
Due to a documented limitation of Microsoft\(aqs
.BR \%setlocale ()
function implementation,
it is not possible to directly select an active locale,
in which the codeset is represented by any multibyte
character sequence with an effective
.B \%MB_CUR_MAX
of more than two bytes.
Prior to \%mingwrt\(hy5.3,
this limitation precludes the use of
.BR \%wcrtomb ()
to interpret any codeset with
.B \%MB_CUR_MAX
greater than two bytes,
(such as
.BR \%UTF\(hy8 ).
From \%mingwrt\(hy5.3 onward,
the MinGW.org implementation of
.BR \%wcrtomb ()
mitigates this limitation by assignment of the codeset
from the
.B \%LC_CTYPE
environment variable,
provided the system default has been previously activated
for the
.B \%LC_CTYPE
locale category;
e.g.\ execution of:
.PP
.RS 4n
.EX
#define _ISOC99_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <locale.h>
#include <limits.h>
#include <wchar.h>

void print_conv( const wchar_t * );

int main()
{
  setlocale( LC_CTYPE, "" );
  putenv( "LC_CTYPE=en_GB.65001" );
  print_conv( L"\eu6c34\eU0001d10b" );
  return 0;
}

void print_conv( const wchar_t *wcs )
{
  wchar_t wch;
  while( (wch = *wcs++) != L'\e0' )
  {
    char mbs[MB_LEN_MAX];
    mbstate_t ps = (mbstate_t)(0);
    size_t n = wcrtomb( mbs, wch, &ps );

    if( (int)(n) > 0 )
    {
      unsigned char *p = (unsigned char *)(mbs);
      printf( "Single wide character: 0x%04X \-\-> %u byte%s",
	  wch, n, (n == 1) ? ":  " : "s: "
	);
      while( n > 0 )
	printf( "0x%02X%c", *p++, (\-\-n == 0) ? '\en' : ':' );
    }
    else if( n == (size_t)(\-1) ) perror( "wcrtomb" );
  }
}
.EE
.RE
.PP
will successfully convert the \fCL"\eu6c34"\fP wide character to its
.B \%UTF\(hy8
equivalent,
resulting in the output:
.PP
.RS 4n
.EX
Single wide character: 0x6C34 \-\-> 3 bytes: 0xE6:0xB0:0xB4
.EE
.RE
.PP
However,
when it then progresses to the \fCL"\eU0001d10b"\fP wide character,
(which
.I should
be represented by a valid
.B \%UTF\(hy16LE
.I surrogate
.IR pair ),
it fails with the diagnostic:
.PP
.RS 4n
.EX
wcrtomb: Invalid or incomplete multibyte or wide character
.EE
.RE
.
.PP
This (possibly unexpected) failure is an unfortunate consequence
of Microsoft\(aqs choice of
.B \%UTF\(hy16LE
as the underlying representation of the
.B \%wchar_t
data type;
this choice makes it impossible for
.I any
\%MS\(hyWindows implementation of
.BR \%wcrtomb ()
to be fully
.B \%ISO\(hyC99
compliant.
To mitigate this non\(hycompliance,
the MinGW implementation of
.BR \%wcrtomb ()
incorporates the following non\(hystandard capabilities:
.RS 2n
.ll -2n
.IP \(bu 2n
When the
.B \%mbstate_t
argument refers to the
.I initial conversion
.IR state ,
and the
.B \%wchar_t
argument represents a
.I high
.IR surrogate ,
then nothing is stored in the conversion buffer,
the
.B \%mbstate_t
reference is updated to indicate pending completion of the
.IR surrogate ,
and the function returns an effective conversion count of
.I zero
bytes.
.
.IP \(bu 2n
When the
.B \%mbstate_t
argument refers to a pending completion of a
.I surrogate
.IR pair ,
and the
.B \%wchar_t
argument represents a
.I low
.IR surrogate ,
then the deferred
.I high surrogate
is combined with the
.I low surrogate
argument,
and the two are converted as a pair;
the resultant conversion is stored in the conversion buffer,
the
.B \%mbstate_t
reference is reset to the
.I initial conversion
.IR state ,
and the function returns the number of bytes
which were stored in the conversion buffer.
.ll +2n
.RE
.
.PP
These capabilities of MinGW\(aqs
.BR \%wcrtomb ()
are certainly non\(hystandard;
nonetheless,
they are required to circumvent non\(hyconformity,
which is imposed by an unfortunate Microsoft design choice,
and it is incumbent upon the caller of
.BR \%wcrtomb (),
on the \%MS\(hyWindows platform,
to make use of them.
The preceding example clearly illustrates how strictly
.B \%ISO\(hyC99
conforming usage will yield incorrect behaviour;
the following illustrates how that example may be adapted,
by incorporation of the above non\(hystandard features,
to achieve correct behaviour:
.PP
.RS 4n
.EX
#define _ISOC99_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <locale.h>
#include <limits.h>
#include <winnls.h>
#include <wchar.h>

void print_conv( const wchar_t * );

int main()
{
  setlocale( LC_CTYPE, "" );
  putenv( "LC_CTYPE=en_GB.65001" );
  print_conv( L"\eu6c34\eU0001d10b" );
  return 0;
}

#define DESC(FMT)  FMT "0x%1$04X --> %2$u byte%3$s"

void print_conv( const wchar_t *wcs )
{
  while( *wcs != L'\e0' )
  {
    wchar_t wch = *wcs;
    char mbs[MB_LEN_MAX];
    mbstate_t ps = (mbstate_t)(0);
    const char *fmt = DESC( "Single wide character: " );
    size_t n = wcrtomb( mbs, wch, &ps );

    if( (n == (size_t)(0)) && IS_HIGH_SURROGATE( wch ) )
    {
      if( (int)(n = wcrtomb( mbs, wcs[1], &ps )) > 0 )
      {
        fmt = DESC( "Surrogate pair: 0x%1$04X:" );
        wcs++;
      }
    }
    if( (int)(n) > 0 )
    {
      unsigned char *p = (unsigned char *)(mbs);
      printf( fmt, wch, n, (n == 1) ? ":  " : "s: ", *wcs );
      while( n > 0 )
        printf( "0x%02X%c", *p++, (\-\-n == 0) ? '\en' : ':' );
    }
    else if( n == (size_t)(\-1) ) perror( "wcrtomb" );
    if( *wcs != L'\e0' ) ++wcs;
  }
}
.EE
.RE
.PP
It may be observed that,
on execution of this modified version of the example,
both the \fCL"\eu6c34"\fP,
and the \fCL"\eU0001d10b"\fP code points are now correctly evaluated,
producing the expected output:
.PP
.RS 2n
.EX
Single wide character: 0x6C34 --> 3 bytes: 0xE6:0xB0:0xB4
Surrogate pair: 0xD834:0xD834 --> 4 bytes: 0xF0:0x9D:0x84:0x8B
.EE
.RE
.
.
.SH SEE ALSO
.BR mbsinit (3),
and
.BR wcsrtombs (3)
.
.
.SH AUTHOR
This manpage was written by \%Keith\ Marshall,
\%<keith@users.osdn.me>,
to document the
.BR \%wcrtomb ()
function as it has been implemented for the MinGW.org Project.
It may be copied, modified and redistributed,
without restriction of copyright,
provided this acknowledgement of contribution by
the original author remains in place.
.
.\" EOF


ChangeLog
2020-07-08  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.4 release.

	* All files (wsl-5.4-release): Tag assigned.

2020-07-07  Keith Marshall  <keith@users.osdn.me>

	Switch to new 5.4-trunk development branch.

	* VERSION.m4 (__VERSION__): Increment it to 5.4
	* All files (5.4-trunk): Bookmark assigned.

2020-07-06  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.3.4 release.

	* All files (wsl-5.3.4-release): Tag assigned.

2020-07-04  Keith Marshall  <keith@users.osdn.me>

	Adjust repository version following WSL-5.3.3 release.

	* VERSION.m4 (__VERSION__): Increment it to 5.3.4

2020-06-03  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.3.3 release.

	* All files (wsl-5.3.3-release): Tag assigned.

2020-06-03  Keith Marshall  <keith@users.osdn.me>

	Discard obsolete distribution files from subdirectories.

	* Makefile.in (dist-subdirs): Add dependency on...
	(dist-subdirs-discard-obsolete): ...this new rule; define it.

2020-06-03  Keith Marshall  <keith@users.osdn.me>

	Revert 2020-05-27 changes relating to issue #40438.

	* VERSION.m4 (LIBMINGWEX): Increment DLL age; this has the effect of
	indicating restored backward compatibility with linmingwex-2.dll

2020-06-03  Keith Marshall  <keith@users.osdn.me>

	Adjust repository version following WSL-5.3.2 release.

	* VERSION.m4 (__VERSION__): Increment it to 5.3.3

2020-05-28  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.3.2 release.

	* All files (wsl-5.3.2-release): Tag assigned.

2020-05-28  Keith Marshall  <keith@users.osdn.me>

	Prepare empty dist directory when required.

	* Makefile.in (dist-prepare): Delete goal; move its commands...
	(dist): ...to here; remove dependency on defective "dist-prepare".

2020-05-27  Keith Marshall  <keith@users.osdn.me>

	Fix libmingwex-n.dll dependencies; cf. Issue #40438.

	* VERSION.m4 (LIBMINGWEX): Increment interface version to 3:0:0.

2020-05-26  Keith Marshall  <keith@users.osdn.me>

	Adjust repository version following WSL-5.3.1 release.

	* VERSION.m4 (__VERSION__): Increment it to 5.3.2

2020-04-30  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.3.1 release.

	* All files (wsl-5.3.1-release): Tag assigned.

2020-03-05  Keith Marshall  <keith@users.osdn.me>

	Adjust repository version following WSL-5.3 release.

	* VERSION.m4 (__VERSION__): Increment it to 5.3.1

2020-03-05  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.3 release.

	* All files (wsl-5.3-release): Tag assigned.

2020-01-13  Keith Marshall  <keith@users.osdn.me>

	Adjust repository version following WSL-5.2.3 release.

	* All files (5.3-trunk): New branch; create, and switch to it.
	* VERSION.m4 (__VERSION__): Increment it to 5.3

2020-01-13  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.2.3 release.

	* All files (wsl-5.2.3-release): Tag assigned.

2019-07-06  Keith Marshall  <keith@users.osdn.me>

	Adjust repository version following WSL-5.2.2 release.

	* VERSION.m4 (__VERSION__): Increment it to 5.2.3

2019-07-06  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.2.2 release.

	* All files (wsl-5.2.2-release): Tag assigned.

2019-02-02  Keith Marshall  <keith@users.osdn.me>

	Add build option for <features.h> sample distribution.

	* configure.ac (lua): Add AC_PATH_PROG check; assign result to...
	(LUA): ...this new AC_SUBST propagated macro definition.

	* Makefile.in (dist): Add dependencies on...
	(dist-prepare, dist-extra): ...these new goals; implement...
	(dist-prepare): ...to initialize an empty staging directory, and...
	(dist-extra): ...incorporate optional distributable components.
	(features.stamp): New optional goal; selected by definition of...
	(FEATURES): ...this user specified macro, nominally defined by...
	(option, options): ...evaluation of command line assignment.
	(features.lua, features.unsupported): New prerequisite goals.
	(WSL_TMPDIR, FEATURES_H_INITIALIZE, FEATURES_CFG, LUA, LUACMD)
	(TODAY, ISSUE): New supporting macros; define them.

2019-01-22  Keith Marshall  <keith@users.osdn.me>

	Adjust repository version following WSL-5.2.1 release.

	* VERSION.m4 (__VERSION__): Increment it to 5.2.2

2019-01-22  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.2.1 release.

	* All files (wsl-5.2.1-release): Tag assigned.

2019-01-22  Keith Marshall  <keith@users.osdn.me>

	Incorporate user-customized <features.h> configuration.

	* Makefile.comm (_mingw.h): Add prerequisite dependency on...
	(features.h): ...this; add rule to create a build-time stub, an empty
	file, yielding the default fall-back configuration.

	* var/features.lua: New direcctory/file; this implements the mingw-get
	post-install action, which is required to integrate management of user
	maintained <features.h> configuration alongside project updates.

2018-12-23  Keith Marshall  <keith@users.osdn.me>

	Adjust repository version following WSL-5.2 release.

	* VERSION.m4 (__VERSION__): Increment it to 5.2.1

2018-12-23  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.2 release.

	* All files (wsl-5.2-release): Tag assigned.

2018-09-04  Keith Marshall  <keith@users.osdn.me>

	Pre-empt default EXEEXT use before its determination.

	* aclocal.m4 (MINGW_AC_NO_EXECUTABLES): Assign an initial assumed
	$ac_exeext value of ".xyz"; it will be updated appropriately, when
	an actual EXEEXT determination has been completed.

2018-09-03  Keith Marshall  <keith@users.osdn.me>

	Adjust repository version following WSL-5.1.1 release.

	* All files (5.2-trunk): New development branch; switch to it.
	* VERSION.m4 (__VERSION__): Increment it to 5.2
	(libmingwex.dll.a): Update ABI version identifier.

2018-09-03  Keith Marshall  <keith@users.osdn.me>

	Prepare and publish MinGW.org WSL-5.1.1 release.

	* All files (wsl-5.1.1-release): Tag assigned.


mingwrt/mingwex/wcrtomb.c
/*
 * wcrtomb.c
 *
 * MinGW.org replacement for the wcrtomb() function; supports use of this
 * function on legacy Windows versions, for which it is not available in the
 * C runtime DLL, and replaces the Microsoft implementation, in those cases
 * where one is available.
 *
 *
 * $Id: wcrtomb.c,v 28b17d1c4eab 2020/07/07 21:02:51 keith $
 *
 * Written by Keith Marshall <keith@users.osdn.me>
 * Copyright (C) 2019, 2020, MinGW.org Project
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice, this permission notice, and the following
 * disclaimer shall be included in all copies or substantial portions of
 * the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OF OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 */
#include "wcharmap.h"

static __mb_inline__ size_t __wcrtomb_internal
( char *restrict mb, wchar_t wc, mbstate_t *ps )
{
  /* Internal implementation of the wcrtomb() function; this is
   * expanded inline, within the public implementation.
   */
  if( *ps != (mbstate_t)(0) )
  {
    /* Microsoft's MBCS implementation does not use shift states;
     * however, it is possible that an immediately preceding call
     * may have stopped with a dangling high surrogate, and thus
     * a restart to complete this, by adding a low surrogate, and
     * converting the pair, may be appropriate.
     */
    if( IS_HIGH_SURROGATE( *ps ) && IS_LOW_SURROGATE( wc ) )
    {
      /* Reassemble the surrogate pair, in a local buffer, and
       * return its conversion, having reset the restart state.
       */
      wchar_t buf[2] = { (wchar_t)(*ps), wc }; *ps = (mbstate_t)(0);
      return __mingw_wctomb_convert( mb, __mingw_wctomb_cur_max(), buf, 2 );
    }
    else
    { /* We expected a low surrogate, but didn't get one; reset
       * the restart state, and abort this conversion.
       */
      *ps = (mbstate_t)(0); return errout( EILSEQ, (size_t)(-1) );
    }
  }
  /* When mb is a NULL pointer, ISO-C99 decrees that the call shall
   * be interpreted as the equivalent of:
   *
   *   wcrtomb( internal_buffer, L'\0', ps );
   *
   * with the encoding of the NUL wchar, preceded by any sequence
   * of bytes needed to restore ps to the initial shift state, being
   * stored in the internal buffer, (and thus, inaccessible to the
   * caller).  Since Microsoft's MBCS encodings do not use shift
   * states, and the encoding for NUL is always a single NUL byte,
   * this becomes the equivalent of returning (size_t)(1).
   */
  if( mb == NULL ) return (size_t)(1);

  /* When mb is not a NULL pointer, then we are obliged to assume
   * that it points to a buffer of at least MB_CUR_MAX bytes, and
   * we may proceed with a normal conversion, (except that, when
   * wc lies in the range reserved for surrogates, we must handle
   * it as a special case.
   */
  if( IS_HIGH_SURROGATE( wc ) )
  { /* A high surrogate is permitted, but it cannot be converted
     * at this time; instead, we simply record that it is present,
     * (subverting ps for this purpose), and move on, without any
     * conversion being performed, and thus storing no converted
     * bytes, in the expection that the next wc passed will be a
     * low surrogate, thus allowing completion of the conversion.
     */
    *ps = (mbstate_t)(wc); return (size_t)(0);
  }
  if( IS_LOW_SURROGATE( wc ) )
    /* A low surrogate, detected here, is an orphan (not paired
     * with a high surrogate from an immediately preceding call);
     * this is not permitted, so report it as invalid.
     */
    return errout( EILSEQ, (size_t)(-1) );

  /* If we get this far, we may proceed with conversion; we return
   * the byte count, and effect of encoding the single wchar which
   * was passed by value in wc.
   */
  return __mingw_wctomb_convert( mb, __mingw_wctomb_cur_max(), &wc, 1 );
}

size_t wcrtomb( char *restrict mb, wchar_t wc, mbstate_t *restrict ps )
{
  /* Implementation of ISO-C99 wcrtomb() function, in libmingwex.a;
   * after storing the effective codeset properties, this returns the
   * result from expansion of the preceding inline implementation.
   */
  (void)(__mingw_wctomb_cur_max_init( __mingw_wctomb_codeset_init() ));
  return __wcrtomb_internal( mb, wc, __mbrtowc_state( ps ) );
}

/* $RCSfile: wcrtomb.c,v $: end of file */
